link-parent/link-ilp-over-http/src/main/java/org/interledger/link/http/auth/JwtHs256BearerTokenSupplier.java

```diff
@@ -0,0 +1,105 @@
+package org.interledger.link.http.auth;
+
+import org.interledger.link.http.OutgoingLinkSettings;
+
+import com.auth0.jwt.JWT;
+import com.auth0.jwt.algorithms.Algorithm;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.cache.RemovalListener;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.Objects;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Supplier;
+
+/**
+ * <p>Defines a {@link Supplier} that provides a JWT that can be used as an ILP-over-HTTP bearer token.</p>
+ *
+ * <p>This implementation attempts to reduce the amount of time the shared-secret is in-memory by preemptively
+ * generating an authentication token that conforms to `JWT_HS_256`, and then wiping the shared-secret from it allocated
+ * memory. Additionally, generated auth tokens have short durations (i.e, they expire), so typing them as Strings is
+ * tolerable for this implementation.</p>
+ */
+public class JwtHs256BearerTokenSupplier implements BearerTokenSupplier {
+
+  private final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+  private final LoadingCache<String, String> ilpOverHttpAuthTokens;
+  private final OutgoingLinkSettings outgoingLinkSettings;
+
+  /**
+   * Required-args Constructor.
+   *
+   * @param sharedSecretBytesSupplier A {@link SharedSecretBytesSupplier} that returns a copy of the shared secret used
+   *                                  to sign JWTs using the HS_256 algorithm.
+   * @param outgoingLinkSettings      A {@link OutgoingLinkSettings} that contains all settings required to construct a
+   *                                  bearer token that can auth an outgoing request.
+   */
+  public JwtHs256BearerTokenSupplier(
+      final SharedSecretBytesSupplier sharedSecretBytesSupplier,
+      final OutgoingLinkSettings outgoingLinkSettings
+  ) {
+    Objects.requireNonNull(sharedSecretBytesSupplier);
+    Objects.requireNonNull(outgoingLinkSettings);
+
+    this.outgoingLinkSettings = Objects.requireNonNull(outgoingLinkSettings);
+
+    ilpOverHttpAuthTokens = CacheBuilder.newBuilder()
+        // There should only ever be 1 or 2 tokens in-memory for a given client instance.
+        .maximumSize(3)
+        // Expire after this duration, which will correspond to the last incoming request from the peer.
+        .expireAfterAccess(this.outgoingLinkSettings.tokenExpiry().orElse(Duration.of(30, ChronoUnit.MINUTES)))
+        .removalListener((RemovalListener<String, String>) notification ->
+            logger.debug("Removing IlpOverHttp AuthToken from Cache for Principal: {}", notification.getKey())
+        )
+        .build(new CacheLoader<String, String>() {
+          public String load(final String authSubject) {
```

Comment:
Maybe you might want to make this as

```java
public String load(@NotNull final String authSubject) {
```

Reply:
Assuming I'm understanding your comment, I consider `@NotNull` annotations in this project to be redundant because we treat null as a bug. Read more here: https://github.com/google/guava/wiki/UsingAndAvoidingNullExplained